README.md
# ebx-timezone-mock

A JavaScript library which mocks the `Date` object and lets you specify a local date/time/timezone in order to enable consistent testing of date/time-related code independently of your local system time settings.

It works by monkey patching the `Date` object and returning results which are adjusted to reflect the effects of both the specified current date/time/timezone (if any) and the current date/time/timezone. In particular, it avoids problems caused by certain JavaScript date/time methods e.g. `getHours` returning results in the user's local system time, so you can write date/time-related unit tests which will work whenever and wherever they are run, without changing timezones causing results to change.

## Installation

### NPM

```sh
$ npm install ebx-timezone-mock
```

## Usage

### Configuration

```js
import TimezoneMock from 'TimezoneMock';

TimezoneMock.set('2018-04-25T12:34:56.000Z');

console.log(new Date()); // Wed Apr 25 2018 12:34:56 GMT+0000 (GMT Standard Time)

TimezoneMock.reset();

console.log(new Date()); // Current system date/time/timezone
```

## Credits

The majority of the code for this project was taken from the [timezone-mock](https://github.com/Jimbly/timezone-mock) project by Jimb Esser,
with some additional code from the [MockDate](https://github.com/boblauer/MockDate) project by Bob Lauer. My contribution has been limited
to gluing these two projects together, and introduction [moment-timezone](https://github.com/moment/moment-timezone/) to solve the problem
of working out the timezone offset for any given date, time and place.


CHANGELOG.md
# Changelog

## Pending

**Bugs:**

* Fix (if necessary) a handful of failing tests - at the moment it isn't clear
  whether these are bugs in this project or in the moment-timezone data

**Environment:**

* Build (compile) using Gulp / Babel
* Implement CI via Travis

## 1.1.4

**Changes:**

* Compiled TimezoneMock.js - mental note, this stuff really needs automating...

## 1.1.3

**Changes:**

* Added `return this` to constructor and all setter methods to allow chaining
  of date methods e.g. `const date = new Date().setYear(2018).setMonth(0);`

## 1.1.2

**Fixes:**

* Switched off DEBUG mode

## 1.1.1

**Fixes:**

* Added compiled version of TimezoneMock.js to root folder
* Updated "main" entry in package.json accordingly

## 1.1.0

**Enhancements:**

* Allow current date/time and timezone to be set
* Use moment-timezone to get timezone offsets for any date/time/timezone

**Environment:**

* Linting implemented using ESLint using Airbnb coding standards
* Formatting implemented using Prettier


test/index.js
import test from 'tape-catch';

import TimezoneMock from '../src/TimezoneMock';

test('Daylight Savings Time Examples', assert => {
  // Tests for Europe/London GMT (UTC+0000)
  assert.test('> Europe/London GMT', assert => {
    assert.test('> getHours', assert => {
      TimezoneMock.set('2018-02-25T12:34:56.000Z', 'Europe/London');
      const now = new Date();
      // console.log(`>>>> now: ${now.getTime()} ${now.toISOString()}`);
      // console.log(`>>>> getHours: ${now.getHours()}`);
      const actual = now.getHours();
      const expected = 12;
      assert.equal(actual, expected, 'In UTC+0000 => hours should be 12');
      TimezoneMock.reset();
      assert.end();
    });
    assert.test('> getUTCHours', assert => {
      TimezoneMock.set('2018-02-25T12:34:56.000Z', 'Europe/London');
      const now = new Date();
      // console.log(`>>>> now: ${now.getTime()} ${now.toISOString()}`);
      // console.log(`>>>> getUTCHours: ${now.getUTCHours()}`);
      let actual = now.getUTCHours();
      let expected = 12;
      assert.equal(actual, expected, 'UTC hours should be as set');
      actual = now.getUTCMinutes();
      expected = 34;
      assert.equal(actual, expected, 'UTC mins should be as set');
      TimezoneMock.reset();
      assert.end();
    });
  });

  // Tests for Europe/London BST (UTC+0100)
  assert.test('> Europe/London BST', assert => {
    assert.test('> getHours', assert => {
      TimezoneMock.set('2018-04-25T12:34:56.000Z', 'Europe/London');
      const now = new Date();
      // console.log(`>>>> now: ${now.getTime()} ${now.toISOString()}`);
      // console.log(`>>>> getHours: ${now.getHours()}`);
      const actual = now.getHours();
      const expected = 13;
      assert.equal(actual, expected, 'In UTC+0100 => hours should be 13');
      TimezoneMock.reset();
      assert.end();
    });
    assert.test('> getUTCHours', assert => {
      TimezoneMock.set('2018-04-25T12:34:56.000Z', 'Europe/London');
      const now = new Date();
      // console.log(`>>>> now: ${now.getTime()} ${now.toISOString()}`);
      // console.log(`>>>> getUTCHours: ${now.getUTCHours()}`);
      let actual = now.getUTCHours();
      let expected = 12;
      assert.equal(actual, expected, 'UTC hours should be as set');
      actual = now.getUTCMinutes();
      expected = 34;
      assert.equal(actual, expected, 'UTC mins should be as set');
      TimezoneMock.reset();
      assert.end();
    });
  });

  // Tests for Asia/Tehran (UTC+0330)
  assert.test('> Asia/Tehran Standard Time', assert => {
    assert.test('> getHours', assert => {
      TimezoneMock.set('2018-02-25T12:34:56.000Z', 'Asia/Tehran');
      const now = new Date();
      // console.log(`>>>> now: ${now.getTime()} ${now.toISOString()}`);
      // console.log(`>>>> getHours: ${now.getHours()}`);
      let actual = now.getHours();
      let expected = 16;
      assert.equal(actual, expected, 'In UTC+0330 => hours should be 16');
      actual = now.getMinutes();
      expected = 4;
      assert.equal(actual, expected, 'In UTC+0330 => mins should be 04');
      TimezoneMock.reset();
      assert.end();
    });
    assert.test('> getUTCHours', assert => {
      TimezoneMock.set('2018-02-25T12:34:56.000Z', 'Asia/Tehran');
      const now = new Date();
      // console.log(`>>>> now: ${now.getTime()} ${now.toISOString()}`);
      // console.log(`>>>> getUTCHours: ${now.getUTCHours()}`);
      let actual = now.getUTCHours();
      let expected = 12;
      assert.equal(actual, expected, 'UTC hours should be as set');
      actual = now.getUTCMinutes();
      expected = 34;
      assert.equal(actual, expected, 'UTC mins should be as set');
      TimezoneMock.reset();
      assert.end();
    });
  });

  // Tests for Asia/Tehran DST (UTC+0430)
  assert.test('> Asia/Tehran Daylight Savings Time', assert => {
    assert.test('> getHours', assert => {
      TimezoneMock.set('2018-04-25T12:34:56.000Z', 'Asia/Tehran');
      const now = new Date();
      // console.log(`>>>> now: ${now.getTime()} ${now.toISOString()}`);
      // console.log(`>>>> getHours: ${now.getHours()}`);
      let actual = now.getHours();
      let expected = 17;
      assert.equal(actual, expected, 'In UTC+0430 => hours should be 17');
      actual = now.getMinutes();
      expected = 4;
      assert.equal(actual, expected, 'In UTC+0430 => mins should be 04');
      TimezoneMock.reset();
      assert.end();
    });
    assert.test('> getUTCHours', assert => {
      TimezoneMock.set('2018-04-25T12:34:56.000Z', 'Asia/Tehran');
      const now = new Date();
      // console.log(`>>>> now: ${now.getTime()} ${now.toISOString()}`);
      // console.log(`>>>> getUTCHours: ${now.getUTCHours()}`);
      let actual = now.getUTCHours();
      let expected = 12;
      assert.equal(actual, expected, 'UTC hours should be as set');
      actual = now.getUTCMinutes();
      expected = 34;
      assert.equal(actual, expected, 'UTC mins should be as set');
      TimezoneMock.reset();
      assert.end();
    });
  });
});


test/setters.js
import test from 'tape-catch';
import * as moment from 'moment-timezone';

import TimezoneMock from '../src/TimezoneMock';

test('Setters', assert => {
  let timezone = moment.tz.guess();
  if (timezone === null) {
    timezone = 'Etc/UTC';
  }

  TimezoneMock.set(new Date().getTime(), timezone);

  const start = '2000-01-01T00:00:00.000Z';
  const orig = new TimezoneMock._Date(start);
  const mock = new Date(start);

  function doit(fn, val) {
    orig[fn](val);
    mock[fn](val);
    // console.log(fn, val, orig, mock);
    let actual = mock.getTime();
    let expected = orig.getTime();
    assert.equal(actual, expected, `${fn} ${val} getTime`);
    actual = mock.getHours();
    expected = orig.getHours();
    assert.equal(actual, expected, `${fn} ${val} getHours`);
    actual = mock.getTimezoneOffset();
    expected = orig.getTimezoneOffset();
    assert.equal(actual, expected, `${fn} ${val} getTimezoneOffset`);
  }

  assert.test('> Specific Dates/Times', assert => {
    doit('setMinutes', 30);
    doit('setFullYear', 2015);
    doit('setHours', 0);
    doit('setMonth', 2);
    doit('setDate', 1);
    doit('setDate', 8);
    doit('setHours', 3);
    doit('setHours', 1);
    doit('setMonth', 9);
    doit('setDate', 1);
    doit('setHours', 1);
    doit('setMonth', 10);
    doit('setHours', 0);
    doit('setHours', 1);
    doit('setHours', 2);
    doit('setHours', 1);
    assert.end();
  });

  function randInt(max) {
    return Math.floor(Math.random() * max);
  }

  assert.test('> Random Dates/Times', assert => {
    for (var ii = 0; ii < 100; ++ii) {
      switch (randInt(3)) {
        case 0:
          doit('setMonth', randInt(12));
          break;
        case 1:
          doit('setDate', randInt(28) + 1);
          break;
        case 2:
          doit('setHours', randInt(24));
          break;
      }
    }
    assert.end();
  });

  TimezoneMock.reset();
});


test/constructors.js
import test from 'tape-catch';
import * as moment from 'moment-timezone';

import TimezoneMock from '../src/TimezoneMock';

test('Constructors', assert => {
  // Test date constructors as used by local timezone mode in node-mysql (local strings)
  let test_str = '2015-01-01 01:23:45.678';
  TimezoneMock.set(new Date().getTime(), 'UTC');
  let actual = new Date(test_str).getTime();
  let expected = 1420104225678 - 8 * 60 * 60 * 1000;
  assert.equal(actual, expected);
  actual = new Date(2015, 0, 1, 1, 23, 45, 678).getTime();
  expected = 1420104225678 - 8 * 60 * 60 * 1000;
  assert.equal(actual, expected);

  TimezoneMock.set(new Date().getTime(), 'America/Los_Angeles');
  actual = new Date(test_str).getTime();
  expected = 1420104225678;
  assert.equal(actual, expected);
  actual = new Date(2015, 0, 1, 1, 23, 45, 678).getTime();
  expected = 1420104225678;
  assert.equal(actual, expected);
  TimezoneMock.set(new Date().getTime(), 'America/New_York');
  actual = new Date(test_str).getTime();
  expected = 1420104225678 - 3 * 60 * 60 * 1000;
  assert.equal(actual, expected);
  actual = new Date(2015, 0, 1, 1, 23, 45, 678).getTime();
  expected = 1420104225678 - 3 * 60 * 60 * 1000;
  assert.equal(actual, expected);

  TimezoneMock.set(new Date().getTime(), 'America/Los_Angeles');
  test_str = '2015-03-08 01:30:00.000'; // right before entering PDT
  actual = new Date(test_str).getTime();
  expected = 1425807000000;
  assert.equal(actual, expected, '**** FAILS ****');
  actual = new Date(test_str).valueOf();
  expected = new Date(test_str).getTime();
  assert.equal(actual, expected);
  actual = new Date(2015, 2, 8, 1, 30, 0, 0).getTime();
  expected = 1425807000000;
  assert.equal(actual, expected, '**** FAILS ****');
  test_str = '2015-03-08 02:30:00.000'; // doesn't exist, ends up 1:30am
  actual = new Date(test_str).getTime();
  expected = 1425807000000;
  assert.equal(actual, expected);
  actual = new Date(2015, 2, 8, 2, 30, 0, 0).getTime();
  expected = 1425807000000;
  assert.equal(actual, expected);
  test_str = '2015-03-08 03:30:00.000'; // in PDT
  actual = new Date(test_str).getTime();
  expected = 1425810600000;
  assert.equal(actual, expected);
  actual = new Date(2015, 2, 8, 3, 30, 0, 0).getTime();
  expected = 1425810600000;
  assert.equal(actual, expected);
  // leaving PDT, JS Date returns 1am PDT, not 1am PST
  // JE: 2017-05-26, Node 6.9.1, Not sure why this was 1am PST before, no changes
  //   to node should have changed how our mock behaves, yet, the mock is still
  //   behaving the same as node, just they are both returning a different value
  //   than the test previously expected, so, updating the test to reflect this.
  test_str = '2014-11-02 01:00:00.000';
  actual = new Date(test_str).getTime();
  expected = 1414915200000;
  assert.equal(actual, expected, '**** FAILS ****');
  actual = new Date(2014, 10, 2, 1, 0, 0, 0).getTime();
  expected = 1414915200000;
  assert.equal(actual, expected, '**** FAILS ****');

  // Testing "local" constructors that look like UTC constructors,
  //   This behavior changed on Node v8.0.0
  actual = new Date('2015-01-01T01:23:45.678').getTime();
  expected = 1420104225678;
  assert.equal(actual, expected);
  actual = new Date('2015-01-01T01:23:45').getTime();
  expected = 1420104225000;
  assert.equal(actual, expected);

  //////////////////////////////////////////////////////////////////////////
  // Test UTC/non-local timezone constructors
  actual = new Date('2017-05-26T17:52:35.869Z').getTime();
  expected = 1495821155869;
  assert.equal(actual, expected);
  actual = new Date('2017-05-26 17:52:35.869 Z').getTime();
  expected = 1495821155869;
  assert.equal(actual, expected);
  actual = new Date('2017-05-26 17:52:35.869 -00:00').getTime();
  expected = 1495821155869;
  assert.equal(actual, expected);
  actual = new Date('2017-05-26 17:52:35.869 +00:00').getTime();
  expected = 1495821155869;
  assert.equal(actual, expected);
  actual = new Date('2017-05-26 18:52:35.869 +01:00').getTime();
  expected = 1495821155869;
  assert.equal(actual, expected);
  actual = new Date('2017-05-26 10:52:35.869 -07:00').getTime();
  expected = 1495821155869;
  assert.equal(actual, expected);
  actual = new Date('2017-05-26 17:52:35 +00:00').getTime();
  expected = 1495821155000;
  assert.equal(actual, expected);
  actual = new Date('2017-05-26 10:52:35 -07:00').getTime();
  expected = 1495821155000;
  assert.equal(actual, expected);

  //////////////////////////////////////////////////////////////////////////
  // Test some generic properties about the date object
  actual = new Date(2017, 5, 1, 0, 0, 0, 0).getTime();
  expected = new Date(2017, 5).getTime();
  assert.equal(actual, expected);

  //////////////////////////////////////////////////////////////////////////
  // Test Brazil timezone oddities
  TimezoneMock.set(new Date().getTime(), 'America/Sao_Paulo');
  test_str = '2017-10-15 00:00:00.000'; // Midnight on this day doesn't exist, jumps to 11PM previous day
  actual = new Date(test_str).getTime();
  expected = 1508032800000;
  assert.equal(actual, expected);
  actual = new Date(2017, 9, 15, 0, 0, 0, 0).getTime();
  expected = 1508032800000;
  assert.equal(actual, expected);

  TimezoneMock.reset();
  assert.end();
});


TimezoneMock.js
'use strict';

var _typeof =
  typeof Symbol === 'function' && typeof Symbol.iterator === 'symbol'
    ? function(obj) {
        return typeof obj;
      }
    : function(obj) {
        return obj &&
          typeof Symbol === 'function' &&
          obj.constructor === Symbol &&
          obj !== Symbol.prototype
          ? 'symbol'
          : typeof obj;
      }; /* eslint func-names:"off" */
/* eslint no-case-declarations:"off" */
/* eslint no-console:"off" */
/* eslint no-global-assign:"off" */
/* eslint no-underscore-dangle:"off" */
/* eslint no-unused-expressions:"off" */
/* eslint prefer-rest-params:"off" */

/* global define */

var _momentTimezone = require('moment-timezone');

var moment = _interopRequireWildcard(_momentTimezone);

function _interopRequireWildcard(obj) {
  if (obj && obj.__esModule) {
    return obj;
  } else {
    var newObj = {};
    if (obj != null) {
      for (var key in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, key))
          newObj[key] = obj[key];
      }
    }
    newObj.default = obj;
    return newObj;
  }
}

(function(name, definition) {
  if (typeof define === 'function' && _typeof(define.amd) === 'object') {
    // AMD
    define(definition);
  } else if (typeof module !== 'undefined' && module.exports) {
    // Node
    module.exports = definition();
  } else {
    // Browser
    this[name] = definition;
  }
})('TimezoneMock', function() {
  // Keep reference to original Date object
  var _Date = Date;

  // Number of milliseconds in an hour
  var HOUR = 60 * 1000;

  // Current date/time and timezone
  var timezone = null;
  var now = null;

  // Regular expressions to match constructor strings
  var REGEX = {
    ISO_8601: /^\d\d\d\d(-\d\d(-\d\d(T\d\d:\d\d:\d\d(\.\d\d\d)?Z)?)?)?$/,
    DATE_WITH_OFFSET: /^\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d(\.\d\d\d)? (Z|(-|\+|)\d\d:\d\d)$/,
    RFC_2822: /^\d\d-\w\w\w-\d\d\d\d \d\d:\d\d:\d\d (\+|-)\d\d\d\d$/,
    LOCAL_DATE: /^\d\d\d\d-\d\d-\d\d[T ]\d\d:\d\d:\d\d(\.\d\d\d)?$/,
  };

  var DEBUG = false;

  function isDefined(value) {
    return typeof value !== 'undefined' && value !== undefined;
  }

  // Constructor function
  function TimezoneMock() {
    DEBUG &&
      console.log(
        '**** TimezoneMock:constructor - args: ' + JSON.stringify(arguments)
      );
    DEBUG && console.log('**** TimezoneMock:constructor - now: ' + now);
    DEBUG &&
      console.log('**** TimezoneMock:constructor - timezone ' + timezone);
    switch (arguments.length) {
      case 0:
        if (now !== null) {
          DEBUG &&
            console.log(
              '**** TimezoneMock:constructor - using existing date ' +
                now.getTime()
            );
          this.date = now;
        } else {
          DEBUG &&
            console.log('**** TimezoneMock:constructor - creating new date');
          this.date = new _Date();
        }
        break;
      case 1:
        var param = arguments[0];
        if (param instanceof TimezoneMock) {
          this.date = new _Date(param.date);
        } else if (typeof param === 'string') {
          if (
            param.match(REGEX.ISO_8601) ||
            param.match(REGEX.DATE_WITH_OFFSET) ||
            param.match(REGEX.RFC_2822)
          ) {
            DEBUG &&
              console.log(
                '**** TimezoneMock:constructor - creating date from string ' +
                  param
              );
            this.date = new _Date(param);
          } else if (param.match(REGEX.LOCAL_DATE)) {
            DEBUG &&
              console.log(
                '**** TimezoneMock:constructor - creating date from local date'
              );
            this.date = new _Date();
            DEBUG &&
              console.log(
                '**** TimezoneMock:constructor - local date created as ' +
                  this.date.toISOString()
              );
            this.fromLocal(new _Date(param.replace(' ', 'T') + 'Z'));
            DEBUG &&
              console.log(
                '**** TimezoneMock:constructor - local date adjusted to ' +
                  this.date.toISOString()
              );
          } else {
            throw new TypeError(
              'Unhandled date format ' +
                param +
                ' passed to TimezoneMock constructor'
            );
          }
        } else if (typeof param === 'number') {
          DEBUG &&
            console.log(
              '**** TimezoneMock:constructor - creating date from number ' +
                param
            );
          this.date = new _Date(param);
        } else {
          throw new TypeError(
            'Unhandled type ' +
              (typeof param === 'undefined' ? 'undefined' : _typeof(param)) +
              ' passed to TimezoneMock constructor'
          );
        }
        break;
      default:
        DEBUG &&
          console.log(
            '**** TimezoneMock:constructor - creating date from multiple arguments'
          );
        var y = isDefined(arguments[0]) ? arguments[0] : 1970;
        var m = isDefined(arguments[1]) ? arguments[1] : 0;
        var d = isDefined(arguments[2]) ? arguments[2] : 1;
        var h = isDefined(arguments[3]) ? arguments[3] : 0;
        var M = isDefined(arguments[4]) ? arguments[4] : 0;
        var s = isDefined(arguments[5]) ? arguments[5] : 0;
        var ms = isDefined(arguments[6]) ? arguments[6] : 0;
        this.date = new _Date(y, m, d, h, M, s, ms);
        this.fromLocal(new _Date(_Date.UTC.apply(null, arguments)));
    }

    return this;
  }

  // Returns the timezone offset for any given date/time in the current timezone
  function tzOffset(timestamp) {
    var result = void 0;
    if (isDefined(timestamp)) {
      result = moment.tz.zone(timezone).parse(timestamp);
      DEBUG &&
        console.log(
          '**** TimezoneMock:tzOffset - offset for timestamp ' +
            timestamp +
            ' in ' +
            timezone +
            ' is ' +
            result
        );
    } else {
      result = moment.tz.zone(timezone).parse(this.date.getTime());
      DEBUG &&
        console.log(
          '**** TimezoneMock:tzOffset - offset for this.date ' +
            this.date.getTime() +
            ' is ' +
            result
        );
    }

    return result;
  }

  TimezoneMock.prototype._tzOffset = tzOffset;

  // Construct methods which use the original Date methods
  function passthrough(fn) {
    TimezoneMock.prototype[fn] = function() {
      var _realdate;

      var realdate = void 0;
      if (this instanceof TimezoneMock) {
        realdate = this.date;
      } else if (this instanceof _Date) {
        realdate = this;
      } else {
        throw new TypeError('Unexpected object type');
      }
      return (_realdate = realdate)[fn].apply(_realdate, arguments);
    };
  }

  // Construct getters which adjust results to take account of local system time
  function localgetter(fn) {
    TimezoneMock.prototype[fn] = function() {
      DEBUG &&
        console.log(
          '**** TimezoneMock.prototype.' +
            fn +
            ' - date: ' +
            this.date.getTime() +
            ' tzoffset: ' +
            this._tzOffset()
        );
      var date = new _Date(
        this.date.getTime() - this._tzOffset(this.date.getTime()) * HOUR
      );
      return date['getUTC' + fn.slice(3)]();
    };
  }

  // Adjusts the date to take account of local system time
  TimezoneMock.prototype.fromLocal = function(date) {
    this.date.setTime(
      date.getTime() +
        this._tzOffset(date.getTime() + this._tzOffset(date.getTime()) * HOUR) *
          HOUR
    );
  };

  // Construct setters which adjust results to take account of local system time
  function localsetter(fn) {
    TimezoneMock.prototype[fn] = function() {
      var date = new _Date(this.date.getTime() - this._tzOffset() * HOUR);
      date['setUTC' + fn.slice(3)].apply(date, arguments);
      this.fromLocal(date);

      return this;
    };
  }

  [
    'getUTCDate',
    'getUTCDay',
    'getUTCFullYear',
    'getUTCHours',
    'getUTCMilliseconds',
    'getUTCMinutes',
    'getUTCMonth',
    'getUTCSeconds',
    'getTime',
    'getTimezoneOffset',
    'setTime',
    'setUTCDate',
    'setUTCFullYear',
    'setUTCHours',
    'setUTCMilliseconds',
    'setUTCMinutes',
    'setUTCMonth',
    'setUTCSeconds',
    'toGMTString',
    'toISOString',
    'toJSON',
    'toUTCString',
    'valueOf',
  ].forEach(passthrough);
  [
    'getDate',
    'getDay',
    'getFullYear',
    'getHours',
    'getMilliseconds',
    'getMinutes',
    'getMonth',
    'getSeconds',
  ].forEach(localgetter);
  [
    'setDate',
    'setFullYear',
    'setHours',
    'setMilliseconds',
    'setMinutes',
    'setMonth',
    'setSeconds',
  ].forEach(localsetter);

  // TimezoneMock.prototype.getTimezoneOffset =
  //   TimezoneMock.prototype._tzOffset;

  TimezoneMock.prototype.toString = function() {
    if (this instanceof _Date) {
      return _Date.prototype.toString.call(this);
    }
    return this.date ? this.date.toISOString() : null; // TBC - need to add GMT+/-xxxx suffix
  };

  TimezoneMock.prototype.toLocaleString = TimezoneMock.prototype.toString;

  TimezoneMock.now = _Date.now;

  TimezoneMock.UTC = _Date.UTC;

  function set(date) {
    var tz =
      arguments.length > 1 && arguments[1] !== undefined
        ? arguments[1]
        : 'Etc/UTC';

    timezone = isDefined(tz) ? tz : moment.tz.guess();
    if (timezone === null) {
      timezone = 'Etc/UTC';
    }

    Date = TimezoneMock;

    now = new _Date(date);

    DEBUG &&
      console.log(
        '**** TimezoneMock:set - now: ' + now + ' timezone: ' + timezone
      );
  }

  function reset() {
    Date = _Date;
  }

  return { set: set, reset: reset, _Date: _Date };
});
